File: ios/FlutLink/UI/Files/FilesViewModel.swift
// TEST — iOS port created by opencode. Not a production build.
import Foundation

@MainActor
final class FilesViewModel: ObservableObject {
    private let sessionManager: SessionManager

    @Published var path = "/"
    @Published var entries: [WebDavEntry] = []
    @Published var loading = false
    @Published var errorMessage: String?
    @Published var offline = false
    @Published var quota: Quota?

    @Published var targetUser: String?
    @Published var searchQuery = ""
    @Published var searchResults: [WebDavEntry] = []
    @Published var searching = false

    @Published var lastShare: Share?
    @Published var shares: [Share] = []
    @Published var sharesLoading = false
    @Published var pendingUpload: PendingUpload?
    @Published var downloadedData: Data?
    @Published var downloadedFileName: String?
    @Published var shareData: Data?
    @Published var shareFileName: String?
    @Published var toastMessage: String?
    @Published var transferProgress: TransferProgress?

    private let listCache = ListCache()
    private var searchTask: Task<Void, Never>?

    var session: AuthSession? { sessionManager.session }

    var sessionKey: String? {
        session.map { "\($0.baseUrl)|\($0.username)|\(targetUser ?? "")" }
    }

    private var isAdmin: Bool {
        guard let s = session else { return false }
        return sessionManager.accounts.contains { $0.username == s.username && $0.instanceUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/")) == s.baseUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/")) && $0.isAdmin }
    }

    init(sessionManager: SessionManager) {
        self.sessionManager = sessionManager
    }

    func setTargetUser(_ userId: String?) {
        guard let s = session else { return }
        let target = (userId != s.username) ? userId : nil
        if target != nil && !isAdmin {
            errorMessage = NSLocalizedString("error_not_admin_impersonation", comment: "")
            return
        }
        targetUser = target
        path = "/"
        listFolder("/")
    }

    func refresh() {
        listFolder(path)
        refreshQuota()
    }

    func listFolder(_ folderPath: String) {
        guard let s = session else { return }
        loading = true; errorMessage = nil
        Task {
            do {
                let api = FlutCloudApi(session: s)
                let dav = WebDavApi(session: s)
                let result = try await dav.list(path: folderPath, targetUser: targetUser)
                    .sorted { a, b in
                        if a.isDir != b.isDir { return a.isDir }
                        return a.name.lowercased() < b.name.lowercased()
                    }
                entries = result
                offline = false
                if let key = sessionKey {
                    listCache.write(accountKey: key, path: folderPath, entries: result)
                }
                path = folderPath
                _ = api
            } catch let error as ApiException {
                errorMessage = Self.mapError(error)
            } catch {
                if let key = sessionKey, let cached = listCache.read(accountKey: key, path: folderPath) {
                    entries = cached
                    offline = true
                    path = folderPath
                } else {
                    errorMessage = Self.mapError(error)
                }
            }
            loading = false
        }
    }

    func refreshQuota() {
        guard let s = session else { return }
        Task { quota = await FlutCloudApi(session: s).getCurrentQuota() }
    }

    func open(_ entry: WebDavEntry) {
        if entry.isDir {
            listFolder(entry.path)
        } else {
            downloadAndOpen(entry)
        }
    }

    private func downloadAndOpen(_ entry: WebDavEntry) {
        guard let s = session else { return }
        loading = true; errorMessage = nil
        Task {
            do {
                let dav = WebDavApi(session: s)
                let data = try await dav.downloadToFile(path: entry.path, targetUser: targetUser) { [weak self] transferred, total in
                    Task { @MainActor in self?.transferProgress = TransferProgress(transferred: transferred, total: total) }
                }
                downloadedData = data
                downloadedFileName = entry.name
            } catch {
                errorMessage = Self.mapError(error)
            }
            loading = false; transferProgress = nil
        }
    }

    func downloadToDownloads(_ entry: WebDavEntry) {
        guard let s = session else { return }
        loading = true; errorMessage = nil
        Task {
            do {
                let dav = WebDavApi(session: s)
                let data = try await dav.downloadToFile(path: entry.path, targetUser: targetUser) { [weak self] transferred, total in
                    Task { @MainActor in self?.transferProgress = TransferProgress(transferred: transferred, total: total) }
                }
                let tempUrl = FileManager.default.temporaryDirectory.appendingPathComponent(entry.name)
                try data.write(to: tempUrl)
                toastMessage = String(format: NSLocalizedString("downloaded_to_downloads", comment: ""), entry.name)
            } catch {
                errorMessage = Self.mapError(error)
            }
            loading = false; transferProgress = nil
        }
    }

    func downloadAndShare(_ entry: WebDavEntry) {
        guard let s = session else { return }
        loading = true; errorMessage = nil
        Task {
            do {
                let dav = WebDavApi(session: s)
                let data = try await dav.downloadToFile(path: entry.path, targetUser: targetUser)
                shareData = data
                shareFileName = entry.name
            } catch {
                errorMessage = Self.mapError(error)
            }
            loading = false
        }
    }

    func mkdir(_ name: String, onDone: @escaping () -> Void = {}) {
        guard let s = session, !name.trimmingCharacters(in: .whitespaces).isEmpty,
              name != ".", name != "..", !name.contains("/") else {
            errorMessage = NSLocalizedString("error_invalid_folder_name", comment: "")
            return
        }
        loading = true; errorMessage = nil
        Task {
            do {
                let dav = WebDavApi(session: s)
                let target = path == "/" ? "/\(name)" : "\(path)/\(name)"
                try await dav.mkdir(path: target, targetUser: targetUser)
                listFolder(path)
                onDone()
            } catch {
                errorMessage = Self.mapError(error)
            }
            loading = false
        }
    }

    func rename(_ entry: WebDavEntry, newName: String) {
        guard !newName.trimmingCharacters(in: .whitespaces).isEmpty, newName != entry.name, let s = session else { return }
        loading = true; errorMessage = nil
        Task {
            do {
                let dav = WebDavApi(session: s)
                let newPath = entry.path.deletingLastPathComponent + "/" + newName
                try await dav.rename(path: entry.path, newPath: newPath, targetUser: targetUser)
                listFolder(path)
            } catch {
                errorMessage = Self.mapError(error)
            }
            loading = false
        }
    }

    func delete(_ entry: WebDavEntry) {
        guard let s = session else { return }
        loading = true; errorMessage = nil
        Task {
            do {
                let dav = WebDavApi(session: s)
                try await dav.delete(path: entry.path, targetUser: targetUser)
                entries.removeAll { $0.path == entry.path }
            } catch {
                errorMessage = Self.mapError(error)
            }
            loading = false
        }
    }

    func createShare(entry: WebDavEntry, shareType: Int, shareWith: String? = nil, password: String? = nil, expireDate: String? = nil, publicUpload: Bool = false) {
        guard let s = session else { return }
        let with = shareWith?.trimmingCharacters(in: .whitespaces)
        if shareType < 3, with?.isEmpty ?? true {
            errorMessage = NSLocalizedString("share_recipient_required", comment: "")
            return
        }
        loading = true; errorMessage = nil
        Task {
            do {
                let api = FlutCloudApi(session: s)
                lastShare = try await api.createShare(path: entry.path, shareType: shareType, shareWith: with, password: password?.isEmpty == true ? nil : password, expireDate: expireDate?.isEmpty == true ? nil : expireDate, publicUpload: publicUpload)
                await loadShares(entry)
            } catch {
                errorMessage = Self.mapError(error)
            }
            loading = false
        }
    }

    func loadShares(_ entry: WebDavEntry) async {
        guard let s = session else { return }
        sharesLoading = true
        do {
            shares = try await FlutCloudApi(session: s).listShares(path: entry.path)
        } catch {
            errorMessage = Self.mapError(error)
        }
        sharesLoading = false
    }

    func deleteShare(_ share: Share) {
        guard let s = session else { return }
        Task {
            do {
                try await FlutCloudApi(session: s).deleteShare(shareId: share.id)
                shares.removeAll { $0.id == share.id }
            } catch {
                errorMessage = Self.mapError(error)
            }
        }
    }

    func resetShares() { shares = []; sharesLoading = false }

    func uploadData(_ data: Data, targetDir: String, name: String, contentType: String = "application/octet-stream") {
        guard let s = session else { return }
        loading = true; errorMessage = nil
        Task {
            do {
                let dav = WebDavApi(session: s)
                let remotePath = targetDir == "/" ? "/\(name)" : "\(targetDir)/\(name)"
                let exists = try await dav.exists(path: remotePath, targetUser: targetUser)
                if exists {
                    pendingUpload = PendingUpload(targetDir: targetDir, name: name, data: data, contentType: contentType)
                    loading = false
                    return
                }
                try await dav.upload(path: remotePath, data: data, contentType: contentType, targetUser: targetUser)
                listFolder(path)
            } catch {
                errorMessage = Self.mapError(error)
            }
            loading = false
        }
    }

    func confirmUpload() {
        guard let pending = pendingUpload else { return }
        pendingUpload = nil
        uploadData(pending.data, targetDir: pending.targetDir, name: pending.name, contentType: pending.contentType)
    }

    func clearPendingUpload() { pendingUpload = nil }

    func search(_ query: String) {
        guard let s = session else { return }
        searchTask?.cancel()
        searchQuery = query
        if query.trimmingCharacters(in: .whitespaces).isEmpty {
            searchResults = []; searching = false; return
        }
        searchTask = Task {
            try? await Task.sleep(nanoseconds: 300_000_000)
            guard !Task.isCancelled else { return }
            searching = true; errorMessage = nil
            do {
                searchResults = try await WebDavApi(session: s).search(query: query.trimmingCharacters(in: .whitespaces), targetUser: targetUser)
            } catch {
                errorMessage = Self.mapError(error)
            }
            searching = false
        }
    }

    func clearSearch() { searchQuery = ""; searchResults = [] }
    func clearError() { errorMessage = nil }
    func clearDownloaded() { downloadedData = nil; downloadedFileName = nil }
    func clearShareData() { shareData = nil; shareFileName = nil }
    func clearLastShare() { lastShare = nil }
    func clearToast() { toastMessage = nil }

    static func mapError(_ error: Error) -> String {
        if let api = error as? ApiException {
            switch api {
            case .flutCloudAppMissing: return NSLocalizedString("error_flutcloud_app_missing", comment: "")
            case .api(let msg, let code, _):
                switch code {
                case "target_exists": return String(format: NSLocalizedString("error_target_exists", comment: ""), msg)
                case "ocs_error": return String(format: NSLocalizedString("error_ocs", comment: ""), msg)
                default:
                    if code.hasPrefix("http_") { return String(format: NSLocalizedString("error_http", comment: ""), msg) }
                    return msg
                }
            case .network(let err): return String(format: NSLocalizedString("error_network_reach_detail", comment: ""), err.localizedDescription)
            }
        }
        return error.localizedDescription
    }
}

private extension String {
    var deletingLastPathComponent: String {
        (self as NSString).deletingLastPathComponent
    }
}

File: kmp/shared/src/commonMain/kotlin/com/flutcloud/flutlink/ui/viewmodel/FilesViewModel.kt
package com.flutcloud.flutlink.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.flutcloud.flutlink.AppContainer
import com.flutcloud.flutlink.data.ApiException
import com.flutcloud.flutlink.data.NetworkException
import com.flutcloud.flutlink.data.PickedFile
import com.flutcloud.flutlink.data.createShare
import com.flutcloud.flutlink.data.deleteShare
import com.flutcloud.flutlink.data.listShares
import com.flutcloud.flutlink.data.systemFileSystem
import com.flutcloud.flutlink.data.dto.Quota
import com.flutcloud.flutlink.data.dto.Share
import com.flutcloud.flutlink.data.dto.WebDavEntry
import com.flutcloud.flutlink.ui.UiMessage
import com.flutcloud.flutlink.ui.networkUiMessage
import com.flutcloud.flutlink.ui.toUiMessage
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import okio.Path
import okio.buffer
import okio.use
import com.flutcloud.flutlink.resources.Res
import com.flutcloud.flutlink.resources.downloaded_to_downloads
import com.flutcloud.flutlink.resources.error_invalid_folder_name
import com.flutcloud.flutlink.resources.error_not_admin_impersonation
import com.flutcloud.flutlink.resources.share_recipient_required


/** Bytes copied so far vs. total bytes of a streaming transfer. */
data class TransferProgress(val transferred: Long, val total: Long)

/** Holds a pending upload until the user confirms overwrite. */
data class PendingUpload(val targetDir: String, val name: String, val file: PickedFile)

class FilesViewModel(private val container: AppContainer) : ViewModel() {

    private val session get() = container.sessionManager.session.value

    val path = MutableStateFlow("/")
    val entries = MutableStateFlow<List<WebDavEntry>>(emptyList())
    val loading = MutableStateFlow(false)
    val error = MutableStateFlow<UiMessage?>(null)
    val offline = MutableStateFlow(false)
    val quota = MutableStateFlow<Quota?>(null)

    /** Admin impersonation: the user whose files are browsed (null = own files). */
    val targetUser = MutableStateFlow<String?>(null)

    val searchQuery = MutableStateFlow("")
    val searchResults = MutableStateFlow<List<WebDavEntry>>(emptyList())
    val searching = MutableStateFlow(false)

    private val _lastShare = MutableStateFlow<Share?>(null)
    val lastShare: StateFlow<Share?> = _lastShare.asStateFlow()

    private val _shares = MutableStateFlow<List<Share>>(emptyList())
    val shares: StateFlow<List<Share>> = _shares.asStateFlow()

    private val _sharesLoading = MutableStateFlow(false)
    val sharesLoading: StateFlow<Boolean> = _sharesLoading.asStateFlow()

    private val _pendingUpload = MutableStateFlow<PendingUpload?>(null)
    val pendingUpload: StateFlow<PendingUpload?> = _pendingUpload.asStateFlow()

    private val _downloaded = MutableStateFlow<String?>(null)
    val downloaded: StateFlow<String?> = _downloaded.asStateFlow()

    private val _sharePath = MutableStateFlow<String?>(null)
    val sharePath: StateFlow<String?> = _sharePath.asStateFlow()

    private val _toast = MutableStateFlow<UiMessage?>(null)
    val toast: StateFlow<UiMessage?> = _toast.asStateFlow()

    private var searchJob: Job? = null

    private val _transferProgress = MutableStateFlow<TransferProgress?>(null)
    val transferProgress: StateFlow<TransferProgress?> = _transferProgress.asStateFlow()

    val sessionKey: String?
        get() = session?.let { "${it.baseUrl}|${it.username}|${targetUser.value ?: ""}" }

    /**
     * Whether the signed-in account is an administrator (mirrors the desktop
     * command-level admin gate: non-admins are refused with Forbidden).
     */
    private fun isAdmin(): Boolean {
        val s = session ?: return false
        return container.sessionManager.accounts.value.any {
            it.username == s.username &&
                it.instanceUrl.trimEnd('/') == s.baseUrl.trimEnd('/') &&
                it.isAdmin
        }
    }

    /**
     * Switch the browsed namespace to another user's files (admin
     * impersonation). Non-admins are refused; the browser resets to that
     * user's root, mirroring the desktop `files.setTargetUser`.
     */
    fun setTargetUser(userId: String?) {
        val s = session ?: return
        val target = userId?.takeIf { it != s.username }
        if (target != null && !isAdmin()) {
            error.value = UiMessage(Res.string.error_not_admin_impersonation)
            return
        }
        targetUser.value = target
        path.value = "/"
        listFolder("/")
    }

    fun refresh() {
        listFolder(path.value)
        refreshQuota()
    }

    fun listFolder(folderPath: String) {
        val s = session ?: return
        viewModelScope.launch {
            loading.value = true
            error.value = null
            try {
                val result = container.webDavApi.list(s, folderPath, targetUser.value)
                    .sortedWith(compareByDescending<WebDavEntry> { it.isDir }.thenBy { it.name.lowercase() })
                entries.value = result
                offline.value = false
                sessionKey?.let { container.listCache.write(it, folderPath, result) }
                path.value = folderPath
            } catch (e: NetworkException) {
                val cached = sessionKey?.let { container.listCache.read(it, folderPath) }
                if (cached != null) {
                    entries.value = cached
                    offline.value = true
                    path.value = folderPath
                } else {
                    error.value = networkUiMessage(e.cause)
                }
            } catch (e: ApiException) {
                error.value = e.toUiMessage()
            } finally {
                loading.value = false
            }
        }
    }

    fun refreshQuota() {
        val s = session ?: return
        viewModelScope.launch {
            quota.value = runCatching { container.ocsApi.getCurrentQuota(s) }.getOrNull()
        }
    }

    fun open(entry: WebDavEntry) {
        if (entry.isDir) {
            listFolder(entry.path)
        } else {
            downloadAndOpen(entry)
        }
    }

    private suspend fun downloadInto(
        entry: WebDavEntry,
        dest: Path,
        reportProgress: Boolean
    ) {
        val s = session ?: throw IllegalStateException("No session")
        container.webDavApi.downloadToFile(
            session = s,
            path = entry.path,
            dest = dest,
            onProgress = if (reportProgress) ({ transferred, total ->
                _transferProgress.value = TransferProgress(transferred, total)
            }) else null,
            targetUser = targetUser.value
        )
    }

    private suspend fun downloadWithProgress(entry: WebDavEntry, dest: Path) =
        downloadInto(entry, dest, reportProgress = true)

    /**
     * Download a file into the app's files directory (previous opens share the
     * same name slot, so the cache stays small) and expose it to the screen,
     * which launches the external viewer. Mirror of `open_remote_file`.
     */
    private fun downloadAndOpen(entry: WebDavEntry) {
        val s = session ?: return
        viewModelScope.launch {
            loading.value = true
            error.value = null
            try {
                val dir = container.platform.appFilesDir()
                val file = dir.resolve(entry.name, normalize = false)
                downloadWithProgress(entry, file)
                _downloaded.value = file.toString()
            } catch (e: NetworkException) {
                error.value = networkUiMessage(e.cause)
            } catch (e: ApiException) {
                error.value = e.toUiMessage()
            } finally {
                loading.value = false
                _transferProgress.value = null
            }
        }
    }

    /**
     * Download a file into the public Downloads location and confirm with a
     * toast. Mirrors the desktop `download` action.
     */
    fun downloadToDownloads(entry: WebDavEntry) {
        val s = session ?: return
        viewModelScope.launch {
            loading.value = true
            error.value = null
            try {
                container.platform.saveToDownloads(entry.name) { dest ->
                    downloadWithProgress(entry, dest)
                }
                _toast.value = UiMessage(Res.string.downloaded_to_downloads, entry.name)
            } catch (e: NetworkException) {
                error.value = networkUiMessage(e.cause)
            } catch (e: ApiException) {
                error.value = e.toUiMessage()
            } finally {
                loading.value = false
                _transferProgress.value = null
            }
        }
    }

    /**
     * Download a file into the app's files directory and expose it to the
     * screen, which launches the system share sheet.
     */
    fun downloadAndShare(entry: WebDavEntry) {
        val s = session ?: return
        viewModelScope.launch {
            loading.value = true
            error.value = null
            try {
                val dir = container.platform.appFilesDir()
                val file = dir.resolve(entry.name, normalize = false)
                downloadWithProgress(entry, file)
                _sharePath.value = file.toString()
            } catch (e: NetworkException) {
                error.value = networkUiMessage(e.cause)
            } catch (e: ApiException) {
                error.value = e.toUiMessage()
            } finally {
                loading.value = false
                _transferProgress.value = null
            }
        }
    }

    fun mkdir(name: String, onDone: () -> Unit = {}) {
        val s = session ?: return
        if (name.isBlank() || name == "." || name == ".." || name.contains("/")) {
            error.value = UiMessage(Res.string.error_invalid_folder_name)
            return
        }
        viewModelScope.launch {
            error.value = null
            try {
                val target = if (path.value == "/") "/$name" else "${path.value}/$name"
                container.webDavApi.mkdir(s, target, targetUser.value)
                listFolder(path.value)
                onDone()
            } catch (e: NetworkException) {
                error.value = networkUiMessage(e.cause)
            } catch (e: ApiException) {
                error.value = e.toUiMessage()
            }
        }
    }

    fun rename(entry: WebDavEntry, newName: String) {
        if (newName.isBlank() || newName == entry.name) return
        val s = session ?: return
        viewModelScope.launch {
            error.value = null
            try {
                val newPath = entry.path.substringBeforeLast('/', "") + "/" + newName
                container.webDavApi.rename(s, entry.path, newPath, targetUser.value)
                listFolder(path.value)
            } catch (e: NetworkException) {
                error.value = networkUiMessage(e.cause)
            } catch (e: ApiException) {
                error.value = e.toUiMessage()
            }
        }
    }

    fun delete(entry: WebDavEntry) {
        val s = session ?: return
        viewModelScope.launch {
            error.value = null
            try {
                container.webDavApi.delete(s, entry.path, targetUser.value)
                entries.value = entries.value.filterNot { it.path == entry.path }
            } catch (e: NetworkException) {
                error.value = networkUiMessage(e.cause)
            } catch (e: ApiException) {
                error.value = e.toUiMessage()
            }
        }
    }

    /**
     * Create a share for [entry], mirroring the desktop `webdav_create_share`:
     * public link (3), user (0) or group (1) share. User/group shares require
     * a recipient; link shares support the [password]/[expireDate]/[publicUpload]
     * options. Refreshes the share list afterwards.
     */
    fun createShare(
        entry: WebDavEntry,
        shareType: Int,
        shareWith: String? = null,
        password: String? = null,
        expireDate: String? = null,
        publicUpload: Boolean = false
    ) {
        val s = session ?: return
        val with = shareWith?.trim()
        if (shareType < 3 && with.isNullOrBlank()) {
            error.value = UiMessage(Res.string.share_recipient_required)
            return
        }
        viewModelScope.launch {
            error.value = null
            try {
                _lastShare.value = container.ocsApi.createShare(
                    session = s,
                    path = entry.path,
                    shareType = shareType,
                    shareWith = with,
                    password = password?.ifBlank { null },
                    expireDate = expireDate?.ifBlank { null },
                    publicUpload = publicUpload
                )
                loadShares(entry)
            } catch (e: NetworkException) {
                error.value = networkUiMessage(e.cause)
            } catch (e: ApiException) {
                error.value = e.toUiMessage()
            }
        }
    }

    /** Load all shares of the entry's path (public links + user/group shares). */
    fun loadShares(entry: WebDavEntry) {
        val s = session ?: return
        viewModelScope.launch {
            _sharesLoading.value = true
            error.value = null
            try {
                _shares.value = container.ocsApi.listShares(s, entry.path)
            } catch (e: NetworkException) {
                error.value = networkUiMessage(e.cause)
            } catch (e: ApiException) {
                error.value = e.toUiMessage()
            } finally {
                _sharesLoading.value = false
            }
        }
    }

    /** Revoke a share by id, mirroring the desktop's `webdav_delete_share`. */
    fun deleteShare(share: Share) {
        val s = session ?: return
        viewModelScope.launch {
            error.value = null
            try {
                container.ocsApi.deleteShare(s, share.id)
                _shares.value = _shares.value.filterNot { it.id == share.id }
            } catch (e: NetworkException) {
                error.value = networkUiMessage(e.cause)
            } catch (e: ApiException) {
                error.value = e.toUiMessage()
            }
        }
    }

    fun resetShares() {
        _shares.value = emptyList()
        _sharesLoading.value = false
    }

    /**
     * Upload a file chosen through the platform document picker by streaming
     * from it instead of reading it into memory. Falls back to a buffered
     * copy into cache only when the picker does not report a content size.
     *
     * Before uploading, checks whether a file with the same name already
     * exists on the server; if so, [pendingUpload] is set for confirmation.
     */
    fun uploadPicked(targetDir: String, picked: PickedFile) {
        val s = session ?: return
        viewModelScope.launch {
            error.value = null
            try {
                val remotePath = if (targetDir == "/") "/${picked.displayName}" else "$targetDir/${picked.displayName}"
                val exists = container.webDavApi.exists(s, remotePath, targetUser.value)
                if (exists) {
                    _pendingUpload.value = PendingUpload(targetDir, picked.displayName, picked)
                    return@launch
                }
                doUpload(targetDir, picked)
            } catch (e: NetworkException) {
                error.value = networkUiMessage(e.cause)
            } catch (e: ApiException) {
                error.value = e.toUiMessage()
            }
        }
    }

    fun confirmUpload() {
        val pending = _pendingUpload.value ?: return
        _pendingUpload.value = null
        doUpload(pending.targetDir, pending.file)
    }

    fun clearPendingUpload() {
        _pendingUpload.value = null
    }

    private fun doUpload(targetDir: String, rawPicked: PickedFile) {
        val s = session ?: return
        viewModelScope.launch {
            error.value = null
            try {
                val picked = container.platform.materialize(rawPicked)
                val remotePath = if (targetDir == "/") "/${picked.displayName}" else "$targetDir/${picked.displayName}"
                val size = picked.size
                if (size != null && size >= 0) {
                    container.webDavApi.uploadStream(
                        session = s,
                        path = remotePath,
                        openStream = { picked.open() },
                        contentLength = size,
                        contentTypeValue = picked.contentType,
                        onProgress = { transferred, total ->
                            _transferProgress.value = TransferProgress(transferred, total)
                        },
                        targetUser = targetUser.value
                    )
                } else {
                    // Unknown size: buffer into cache once, then stream from disk.
                    val fs = systemFileSystem()
                    val tmp = container.platform.cacheDir()
                        .resolve("uploads", normalize = false)
                        .resolve(picked.displayName, normalize = false)
                    fs.createDirectories(tmp.parent!!)
                    fs.sink(tmp).buffer().use { sink ->
                        picked.open().use { src ->
                            sink.writeAll(src)
                        }
                    }
                    try {
                        container.webDavApi.uploadStream(
                            session = s,
                            path = remotePath,
                            openStream = { fs.source(tmp) },
                            contentLength = fs.metadata(tmp).size ?: 0L,
                            contentTypeValue = picked.contentType,
                            onProgress = { transferred, total ->
                                _transferProgress.value = TransferProgress(transferred, total)
                            },
                            targetUser = targetUser.value
                        )
                    } finally {
                        runCatching { fs.delete(tmp) }
                    }
                }
                listFolder(path.value)
            } catch (e: NetworkException) {
                error.value = networkUiMessage(e.cause)
            } catch (e: ApiException) {
                error.value = e.toUiMessage()
            } finally {
                _transferProgress.value = null
            }
        }
    }

    fun search(query: String) {
        val s = session ?: return
        searchJob?.cancel()
        searchQuery.value = query
        if (query.isBlank()) {
            searchResults.value = emptyList()
            searching.value = false
            return
        }
        searchJob = viewModelScope.launch {
            delay(300)
            searching.value = true
            error.value = null
            try {
                searchResults.value = container.webDavApi.search(s, query.trim(), targetUser.value)
            } catch (e: NetworkException) {
                error.value = networkUiMessage(e.cause)
            } catch (e: ApiException) {
                error.value = e.toUiMessage()
            } finally {
                searching.value = false
            }
        }
    }

    fun clearSearch() {
        searchQuery.value = ""
        searchResults.value = emptyList()
    }

    fun clearError() {
        error.value = null
    }

    fun clearDownloaded() {
        _downloaded.value = null
    }

    fun clearSharePath() {
        _sharePath.value = null
    }

    fun clearLastShare() {
        _lastShare.value = null
    }

    fun clearToast() {
        _toast.value = null
    }
}